pkg/storage/tsdb/expanded_postings_cache.go

```diff
@@ -0,0 +1,422 @@
+	Ttl      time.Duration `yaml:"ttl"`
+	Enabled  bool          `yaml:"enabled"`
+}
+
+func (cfg *TSDBPostingsCacheConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	cfg.Head.RegisterFlagsWithPrefix(prefix, "head", f)
+	cfg.Blocks.RegisterFlagsWithPrefix(prefix, "block", f)
+}
+
+// RegisterFlagsWithPrefix adds the flags required to config this to the given FlagSet
+func (cfg *PostingsCacheConfig) RegisterFlagsWithPrefix(prefix, block string, f *flag.FlagSet) {
+	f.Int64Var(&cfg.MaxBytes, prefix+"expanded_postings_cache."+block+".max-bytes", 10*1024*1024, "Max bytes for postings cache")
+	f.IntVar(&cfg.MaxItems, prefix+"expanded_postings_cache."+block+".max-items", 10000, "Max items for postings cache")
+	f.DurationVar(&cfg.Ttl, prefix+"expanded_postings_cache."+block+".ttl", 10*time.Minute, "TTL for postings cache")
+	f.BoolVar(&cfg.Enabled, prefix+"expanded_postings_cache."+block+".enabled", false, "Whether the postings cache is enabled or not")
+}
+
+type ExpandedPostingsCache interface {
+	PostingsForMatchers(ctx context.Context, blockID ulid.ULID, ix tsdb.IndexReader, ms ...*labels.Matcher) (index.Postings, error)
+	ExpireSeries(metric labels.Labels)
+}
+
+type BlocksPostingsForMatchersCache struct {
+	strippedLock []sync.RWMutex
+
+	headCache   *fifoCache[[]storage.SeriesRef]
+	blocksCache *fifoCache[[]storage.SeriesRef]
+
+	headSeedByMetricName    []int
+	postingsForMatchersFunc func(ctx context.Context, ix tsdb.IndexReader, ms ...*labels.Matcher) (index.Postings, error)
+	timeNow                 func() time.Time
+
+	metrics *ExpandedPostingsCacheMetrics
+}
+
+func NewBlocksPostingsForMatchersCache(cfg TSDBPostingsCacheConfig, metrics *ExpandedPostingsCacheMetrics) ExpandedPostingsCache {
+	if cfg.PostingsForMatchers == nil {
+		cfg.PostingsForMatchers = tsdb.PostingsForMatchers
+	}
+
+	if cfg.timeNow == nil {
+		cfg.timeNow = time.Now
+	}
+
+	return &BlocksPostingsForMatchersCache{
+		headCache:               newFifoCache[[]storage.SeriesRef](cfg.Head, "head", metrics, cfg.timeNow),
+		blocksCache:             newFifoCache[[]storage.SeriesRef](cfg.Blocks, "block", metrics, cfg.timeNow),
+		headSeedByMetricName:    make([]int, seedArraySize),
+		strippedLock:            make([]sync.RWMutex, numOfSeedsStripes),
+		postingsForMatchersFunc: cfg.PostingsForMatchers,
+		timeNow:                 cfg.timeNow,
+		metrics:                 metrics,
+	}
+}
+
+func (c *BlocksPostingsForMatchersCache) ExpireSeries(metric labels.Labels) {
+	metricName, err := extract.MetricNameFromLabels(metric)
+	if err != nil {
+		return
+	}
+
+	h := MemHashString(metricName)
+	i := h % uint64(len(c.headSeedByMetricName))
+	l := h % uint64(len(c.strippedLock))
+	c.strippedLock[l].Lock()
+	defer c.strippedLock[l].Unlock()
+	c.headSeedByMetricName[i]++
+}
+
+func (c *BlocksPostingsForMatchersCache) PostingsForMatchers(ctx context.Context, blockID ulid.ULID, ix tsdb.IndexReader, ms ...*labels.Matcher) (index.Postings, error) {
```

Comment:
Are we missing unit test for this func?

Reply:
I dont think so, both `TestExpandedPostingsCacheFuzz` and `TestExpendedPostingsCache` tests this function.

Reply:
Current Coverage:

```
github.com/cortexproject/cortex/pkg/storage/tsdb/expanded_postings_cache.go:127:		ExpireSeries					88.9%
github.com/cortexproject/cortex/pkg/storage/tsdb/expanded_postings_cache.go:141:		PostingsForMatchers				100.0%
github.com/cortexproject/cortex/pkg/storage/tsdb/expanded_postings_cache.go:145:		fetchPostings					95.8%
```